letsencrypt/client/network2.py

```diff
@@ -0,0 +1,480 @@
+            Authorization Resources (`.AuthorizationResource`) as
+            present in the responses from server, and in the same order
+            as the input ``authzrs``.
+        :rtype: `tuple`
+
+        """
+        # priority queue with datetime (based od Retry-After) as key,
+        # and original Authorization Resource as value
+        waiting = [(datetime.datetime.now(), authzr) for authzr in authzrs]
+        # mapping between original Authorization Resource and the most
+        # recently updated one
+        updated = dict((authzr, authzr) for authzr in authzrs)
+
+        while waiting:
+            # find the smallest Retry-After, and sleep if necessary
+            when, authzr = heapq.heappop(waiting)
+            now = datetime.datetime.now()
+            if when > now:
+                seconds = (when - now).seconds
+                logging.debug('Sleeping for %d seconds', seconds)
+                time.sleep(seconds)
+
+            # Note that we poll with the latest updated Authorization
+            # URI, which might have a different URI than initial one
+            updated_authzr, response = self.poll(updated[authzr])
+            updated[authzr] = updated_authzr
+
+            if updated_authzr.body.status != messages2.STATUS_VALID:
+                # push back to the priority queue, with updated retry_after
+                heapq.heappush(waiting, (self.retry_after(
+                    response, default=mintime), authzr))
+
+        updated_authzrs = tuple(updated[authzr] for authzr in authzrs)
+        return self.request_issuance(csr, updated_authzrs), updated_authzrs
+
+    def _get_cert(self, uri):
+        content_type = self.DER_CONTENT_TYPE  # TODO: make it a param
+        response = self._get(uri, headers={'Accept': content_type},
+                             content_type=content_type)
+        return response, jose.ComparableX509(
+            M2Crypto.X509.load_cert_der_string(response.content))
+
+    def check_cert(self, certr):
+        """Check for new cert.
+
+        :param certr: Certificate Resource
+        :type certr: `.CertificateResource`
+
+        :returns: Updated Certificate Resource.
+        :rtype: `.CertificateResource`
+
+        """
+        # TODO: acme-spec 5.1 table action should be renamed to
+        # "refresh cert", and this method integrated with self.refresh
+        response, cert = self._get_cert(certr.uri)
+        if 'Location' not in response.headers:
+            raise errors.NetworkError('Location header missing')
+        if response.headers['Location'] != certr.uri:
+            raise errors.UnexpectedUpdate(response.text)
+        return certr.update(body=cert)
+
+    def refresh(self, certr):
+        """Refresh certificate.
+
+        :param certr: Certificate Resource
+        :type certr: `.CertificateResource`
+
+        :returns: Updated Certificate Resource.
+        :rtype: `.CertificateResource`
+
+        """
+        return self.check_cert(certr)
```

Comment:
According to section 5.7 this can also return a 403 (I am fine with it being a TODO)